src/fast_layers.jl

```diff
@@ -24,84 +24,160 @@ initial_params(c::FastChain) = vcat(initial_params.(c.layers)...)
 
 """
 FastDense(in,out,activation=identity;
-          bias = true ,initW = Flux.glorot_uniform, initb = Flux.zeros32)
+          bias = true, precache = false ,initW = Flux.glorot_uniform, initb = Flux.zeros32)
 
 A Dense layer `activation.(W*x + b)` with input size `in` and output size `out`.
 The `activation` function defaults to `identity`, meaning the layer is an affine
 function. Initial parameters are taken to match `Flux.Dense`. 'bias' represents b in
-the layer and it defaults to true.
+the layer and it defaults to true.'precache' is used to preallocate memory for the
+intermediate variables calculated during each pass. This avoids heap allocations
+in each pass which would otherwise slow down the computation, it defaults to false.
 
 Note that this function has specializations on `tanh` for a slightly faster
 adjoint with Zygote.
 """
-struct FastDense{F,F2} <: FastLayer
+struct FastDense{F,F2,C} <: FastLayer
   out::Int
   in::Int
   σ::F
   initial_params::F2
+  cache :: C
   bias::Bool
+  numcols::Int64
   function FastDense(in::Integer, out::Integer, σ = identity;
-                 bias = true, initW = Flux.glorot_uniform, initb = Flux.zeros32)
+                 bias = true, numcols=1, precache=false, initW = Flux.glorot_uniform, initb = Flux.zeros32)
     temp = ((bias == false) ? vcat(vec(initW(out, in))) : vcat(vec(initW(out, in)),initb(out)))
     initial_params() = temp
-    new{typeof(σ),typeof(initial_params)}(out,in,σ,initial_params,bias)
+    if precache == true
+      cache = (
+        W = zeros(out, in),
+        y = zeros(out, numcols),
+        r = zeros(out, numcols),
+        zbar = zeros(out, numcols),
+        Wbar = zeros(out, in),
+        xbar = zeros(in, numcols),
+        pbar = if bias == true
+            zeros((out*in)+out)
+          else
+            zeros(out*in)
+          end
+        )
+      else
+        cache = nothing
+      end
+      new{typeof(σ), typeof(initial_params), typeof(cache)}(out,in,σ,initial_params,cache,bias,numcols)
+      # new{typeof(σ),typeof(initial_params)}(out,in,σ,initial_params,bias)
   end
 end
 
 # (f::FastDense)(x,p) = f.σ.(reshape(uview(p,1:(f.out*f.in)),f.out,f.in)*x .+ uview(p,(f.out*f.in+1):lastindex(p)))
 (f::FastDense)(x,p) = ((f.bias == true) ? (f.σ.(reshape(p[1:(f.out*f.in)],f.out,f.in)*x .+ p[(f.out*f.in+1):end])) : (f.σ.(reshape(p[1:(f.out*f.in)],f.out,f.in)*x)))
 
 ZygoteRules.@adjoint function (f::FastDense)(x,p)

```

Comment:
Make this two separate dispatches, one for `x::AbstractVector` and one for `x::AbstractMatrix`. That will make it a lot simpler.